react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+2. If you've passed `memoizedHandleClick` and the `ButtonComponent` has a memo. It will not re-render. Referential equality check passes (previous prop is *equal* to the current prop).
+
+This works with all values that will be passed as a prop. You might see it being used frequently with the Context API.
+
+~~~jsx
+const value = useMemo(
+  () => ({ someState, someFunction }),
+  [someState, someFunction]
+);
+
+return <Context.Provider value={value}>{children}</Context.Provider>;
+~~~
+
+### The useCallback hook
+
+The `useCallback` hook provides another way to memoize a value, not just **any** value like `useMemo`. It can only memoize a function. Did you see the previous snippet that we have with memoizing a function reference with `useMemo`?
+
+~~~jsx
+const handleClick = useMemo(
+  () => () => setCount((prevState) => prevState + 1),
+  []
+);
+// or
+const memoizedHandleClick = useMemo(() => handleClick, []);
+~~~
+
+With `useCallback`, we don't need to do that. It's specifically made for functions:
+
+~~~jsx
+import { useCallback } from "react";
+
+// Inside a component
+// Without useCallback
+const handleClick = () => setCount((prevState) => prevState + 1);
+// With useCallback
+const handleClick = useCallback(
+  () => setCount((prevState) => prevState + 1),
+  []
+);
+// or
+const memoizedHandleClick = useCallback(handleClick, []);
+~~~
+
+<a name="usememo-or-usecallback"></a>
+Yay, there's only one arrow function, and it's simpler to read. There's nothing extra to `useCallback` other than it only memoizes functions. So the main difference between `useMemo` and `useCallback` is just the type of value it returns.
+
+Which one should I use, then? Use `useMemo` for *any* value types, and use `useCallback` specifically for functions. At the end of the day, they both do similar things with a tiny difference, so use whatever you prefer.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. The article [When to useMemo and useCallback](https://kentcdodds.com/blog/usememo-and-usecallback) by Kent C. Dodds further introduces more examples of when to use `useMemo` and `useCallback` and why you shouldn't bother using it.
+2. We've only learned about a basic implementation of the `useRef` hook. For more examples about its usage and why we should be wary of using the hook (more on the links they provided in the guide), check out the [interactive guide of the React documentation](https://react.dev/reference/react/useRef) for `useRef`. Try your best to complete all the challenges!
+3. The article [useRef instead of querySelector in React](https://meje.dev/blog/useref-not-queryselector) by Caleb Olojo briefly tells some unexpected behaviors when trying to manipulate the DOM directly with DOM manipulation methods and why we should prefer `useRef` over other DOM manipulation methods like `querySelector`. Check it out!
+
```

Comment:
Since the example with DOM manipulation was addressed by the lesson, would an example on the use of refs for storing interval/timeout ids be good in the assignements?
Dan Abhramov has a great article on this: https://overreacted.io/making-setinterval-declarative-with-react-hooks/

Reply:
That's a great addition, thanks 👍

Reply:
Love this article